db/db_log_iter_test.cc

```diff
@@ -179,6 +179,8 @@ TEST_F(DBTestXactLogIterator, TransactionLogIteratorCheckWhenArchive) {
 
     delete cf;
   } while (ChangeCompactOptions());
+  ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
+  Close();
```

Comment:
Agree with the analysis and this change would indeed make the issue go away; however, the test completing before the callback is called means that `ExpectRecords`, which is the heart of the test, is not called in time at least in one configuration. I think what we might want to do instead is to set up a dependency to ensure that `PurgeObsoleteFiles` actually completes before we move on, like in https://github.com/facebook/rocksdb/pull/11537

Reply:
I tried a bit to understand/reproduce the async purge, but hit my time limit.

The sync point is hit several times per iteration, so I would guess that it's always going to be hit at least once by the end of close.